### ISSUE TITLE
too much API breakage - mandatory use of semantic versioning

### ISSUE THREAD

**Issue by conradsnicta** [HDFGroup/hdf5#2]
HDF5 v1.12 has changed so much since the initial release of HDF5 that it's more appropriate to call it HDF**6** at this stage.

The breaking API changes between minor HDF5 releases (eg. 1.8, 1.10, 1.12) are rather disconcerting.  While there are various pre-processor tricks used as "workarounds", they are just brittle hacks in the end. Stuff like `H5_USE_110_API` is just papering over the unwarranted breakage.

The point of HDF is to be a boring abstraction for storage functionality, where by "boring" I mean stable. HDF5 is definitely not boring, which decreases its value and utility.

The use of [semantic versioning](https://semver.org/) to label HDF releases would help in this regard immensely. Minor releases such as v1.12 should not be introducing any kind of API breaks. Such breaks are much better communicated by bumping the major version number, for example HDF6, HDF7, etc.

The current approach of API breakage between minor releases is causing all sorts of issues. Examples:
* https://gitlab.com/conradsnicta/armadillo-code/-/issues/155
* https://github.com/Homebrew/homebrew-core/issues/52003
* https://github.com/mlpack/mlpack/pull/2328

TLDR: I would suggest that the HDF5 developers be made explicitly aware that public API breaks in HDF5 releases are not exactly conducive to productivity. It would be useful to _properly_ communicate such changes via an increase in the major version number, for example HDF**6**. The current practice of updating "HDF5" with incompatible changes is overall a negative net contribution.

**Comment by ArchangeGabriel** [comment]
HDF6 is not a correct major version number change, HDF5 2.x would be. AFAIK the 5 in HDF5 relates to the storage specification, and that is stable AFAICT. What is changing here is the API of the official library for accessing this file format, and this library has version 1.12.x when it should likely be 2.x.y.

**Comment by scivision** [comment]
Also msys2 / h5py is broken with hdf5 1.12

**Comment by conradsnicta** [comment]
> HDF6 is not a correct major version number change, HDF5 2.x would be. AFAIK the 5 in HDF5 relates to the storage specification, and that is stable AFAICT. What is changing here is the API of the official library for accessing this file format, and this library has version 1.12.x when it should likely be 2.x.y.

This is probably a good trade-off.

**Comment by tbeu** [comment]
Same holds for [libmatio](https://github.com/tbeu/matio) where I needed to make the appropriate changes to support HDF5 v1.8x, v1.10.x and v1.12.0.

**Comment by derobins** [comment]
Proper semantic versioning will begin with HDF5 2.0.0, which will probably release sometime in 2023, after 1.14.0 releases.

**Comment by conradsnicta** [comment]
The eventual use of semantic versioning with HDF5 2.0.0 is certainly welcome, though it should have been applied much earlier.

Why not simply skip 1.14 and go directly to 2.0?  The sooner HDF5 gets out of this mess, the better, saving everybody a metric ton of headaches.
